### DISCUSSION TITLE
[Preview] Allow public access to the new BuildCargoData method for assets.

### DISCUSSION THREAD

**Discussion by DanielWillett** [SmartlyDressedGames/Unturned-3.x-Community#4808] [Ideas]
The new BuildCargoData could be used for exporting Assets to any format for various reasons which has been a bit cumbersome to do in the past.

If possible would you be willing to abstract CargoBuilder/CargoDeclaration behind an interface and expose that method?

ps. thank you so much for redoing the explode method (where it has to overlap the center of a vehicle to damage it)

**Comment by SDGNelson** [comment]
I completely agree that converting asset data would be really useful. I've had a few related ideas in my head like:
- Development option to load dats "with metadata" like ordering in the file, whitespace, comments, etc., in the direction of automatic conversions and modifying the dat files.
- Making PopulateAsset bidirectional. Rather than ``x = Read("x")`` more like ``Serialize("x", ref x)`` so that a different implementation can write all the properties read by it.

The Cargo data export, conversely, is @MoltonMontro's project to automate data population on the wiki, and is very work-in-progress so the format and organization is likely to change. This is why it's internal for the moment.

For your data conversion perhaps it could be built externally to the asset classes? What format are you aiming to convert to?

Thanks for the P.S. - good point that small explosion damage against vehicles like frag ammo should work much better now! :)

**Reply by DanielWillett** [reply]
Yeah, i mean it could always just be read from the file after exporting, just thought it would be nice to be able to have it work with any format, like json, databases, etc. A while back i tried to store assets in an SQL database but getting all properties and keeping it up to date would have been quite a headache.

The multidirectional serialization with extra metadata sounds like it would be a good way to handle assets in the future. Could maybe be done similar to how MSBuild handles project files, where it just replaces properties without touching the rest of the file.

**Comment by MoltonMontro** [comment]
Thanks, Nelson!

With that in mind, even _without_ being exposed, players can export Cargo data from the Unturned Wiki. You'd have access to any vanilla/curated items added to those Cargo tables, but you'd have to rely on the wiki for the information.

Using this placeholder "Animals" table as an example: https://unturned.wiki.gg/wiki/User:MoltonMontro/CargoExport?oldid=42411

Cargo data can also be generated for mods. Although these wouldn't be added to our official wiki, anyone hosting a MediaWiki site (e.g., on FANDOM or wiki.gg) could add their own Cargo data. And then that data could be exported as well! :)
